src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,49 +1,172 @@
+            elementData = grow();
+        }
+        elementData[s] = value;
+        size = s + 1;
     }
 
     @Override
     public void add(T value, int index) {
-
+        rangeCheckForAdd(index);
+        final int s;
+        Object[] elementData;
+        if ((s = size) == (elementData = this.elementData).length) {
+            elementData = grow();
+        }
+        System.arraycopy(elementData, index,
+                elementData, index + 1,
+                s - index);
+        elementData[index] = value;
+        size = s + 1;
     }
 
     @Override
     public void addAll(List<T> list) {
+        Object[] arrayForAdd = new Object[list.size()];
+        for (int i = 0; i < list.size(); i++) {
+            arrayForAdd[i] = list.get(i);
+        }
+        int addLength = arrayForAdd.length;
+        Object[] elementData;
+        final int s;
+        if (addLength > (elementData = this.elementData).length - (s = size)) {
+            elementData = grow(s + addLength);
+        }
+        System.arraycopy(arrayForAdd, 0, elementData, s, addLength);
+        size = s + addLength;
+    }
 
+    T elementData(int index) {
+        return (T) elementData[index];
     }
 
     @Override
     public T get(int index) {
-        return null;
+        rangeCheck(index);
+        return elementData(index);
+    }
+
+    private Object[] grow(int minCapacity) {
+        return elementData = Arrays.copyOf(elementData,
+                newCapacity(minCapacity));
+    }
+
+    private Object[] grow() {
+        return grow(size + 1);
+    }
+
+    private static int hugeCapacity(int minCapacity) {
+        if (minCapacity < 0) {
+            throw new OutOfMemoryError();
+        }
+        return (minCapacity > MAX_ARRAY_SIZE)
+                ? Integer.MAX_VALUE
+                : MAX_ARRAY_SIZE;
     }
 
     @Override
-    public void set(T value, int index) {
+    public boolean isEmpty() {
+        return size == 0;
+    }
+
+    private int newCapacity(int minCapacity) {
+        int oldCapacity = elementData.length;
+        int newCapacity = oldCapacity + (oldCapacity >> 1);
+        if (newCapacity - minCapacity <= 0) {
+            if (elementData == DEFAULTCAPACITY_EMPTY_ELEMENTDATA) {
+                return Math.max(DEFAULT_CAPACITY, minCapacity);
+            }
+            if (minCapacity < 0) {
+                throw new OutOfMemoryError();
+            }
+            return minCapacity;
+        }
+        return (newCapacity - MAX_ARRAY_SIZE <= 0)
+                ? newCapacity
+                : hugeCapacity(minCapacity);
+    }
+
+    private void rangeCheckForAdd(int index) {
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException(MESSAGE_INDEX_ERROR);
+        }
+    }
 
+    private void rangeCheck(int index) {
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException(MESSAGE_INDEX_ERROR);
+        }
     }
```

Comment:
these methods are the same, so let's create one common method witch will get two parameters: int index and int maxSize, and will check the index